cmd/libsnap-confine-private/snap-dir.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include "snap-dir.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "snap.h"  // For SC_SNAP_MOUNT_DIR_UNSUPPORTED
+
+static const char *_snap_mount_dir = NULL;
+
+const char *sc_snap_mount_dir(sc_error **errorp) {
+    sc_error *err = NULL;
+
+    if (_snap_mount_dir == NULL) {
+        err = sc_error_init_api_misuse("sc_probe_snap_mount_dir was not called yet");
+    }
+
+    sc_error_forward(errorp, err);
+    return _snap_mount_dir;
+}
+
+void sc_probe_snap_mount_dir(int root_fd, sc_error **errorp) {
+    char target[PATH_MAX] = {0};
+    struct stat sb;
+    ssize_t n = 0;
+    sc_error *err = NULL;
+
+    const char *const probe_dir =
+        root_fd == AT_FDCWD ? "/proc/1/root" SC_CANONICAL_SNAP_MOUNT_DIR : "proc/1/root" SC_CANONICAL_SNAP_MOUNT_DIR;
```

Comment:
I appreciate the compact syntax byt `if (root_fd == AT_FDCWD) { .... } ` would still be more explicit and easier to follow given the amount of string concatenation that's happening here. The whole thing could use a comment.

Reply:
I wanted to keep the `const char *const *probe_dir` as const as I can. I re-wrote it with different spacing and extensive comments. Have a look please.